### ISSUE TITLE
AD FS 3.0 and OAuth

### ISSUE THREAD

**Issue by durandt** [xamarin/Xamarin.Auth#92]
Hi,

Did anyone get Xamarin.Auth to work with AD FS using OAuth?

One thing that strikes me is that Xamarin.Auth uses the "scope" parameter, where it seems that AD FS would rather have a "resource" parameter. (see http://blog.scottlogic.com/2015/03/09/OAUTH2-Authentication-with-ADFS-3.0.html)

```
 I found that most OAUTH2 libraries expect to pass a scope but not a resource parameter.
```

Regards,

Thibault D

**Comment by rgccomtec** [comment]
Yes we did use Xamarin with ADFS by extending OAuth2Authenticator. I've blogged it below. Let me know if I can help.

http://comtecblog.azurewebsites.net/2016/08/03/xamarin-and-oauth2-with-adfs/

Rahul Gangwar

**Comment by MarcBruins** [comment]
Hi @rgccomtec, your link is down. Would be nice to see the blogpost however!

**Comment by Diongerritzen** [comment]
@MarcBruins, I managed to find the blogpost: http://blogs.comtecinfo.com/xamarin-and-oauth2-with-adfs/

**Comment by rg42** [comment]
I'm sorry @MarcBruins, the blogpost was moved to official site of company. 
Thanks @Diongerritzen for providing the correct link.
Feel free to contact me directly at RG@outlook.in

**Comment by charithjj** [comment]
Unfortunately the second link also broken. Would you please provide the current link?

**Comment by durandt** [comment]
Hi, 

It's pretty straightforward to use OAuth 2 with AD FS. The most misleading thing is that AD FS does not follow the OAuth 2 RFC, like i mentioned in my first post. The `scope` parameter is named `resource`, for example. 

I found that the most useful documentation about AD FS's OAuth [can be found here](https://github.com/nordvall/TokenClient/wiki/OAuth-2-Authorization-Code-grant-in-ADFS). 

I've also made a [fork](https://github.com/durandt/ADFSOAuth-xamarin-sample) of Xamarin.Auth showing what changes needed to be done to make it work with AD FS's OAuth and you can find it [here](https://github.com/durandt/ADFSOAuth-xamarin-sample). 

I will advise you against using it though. Xamarin.Auth is not maintained and it's original code quality is bad. After I made my sample work with Xamarin.Auth, I wrote my own based on Xamarin.Forms' `WebView` and used native renderers to catch some events, like redirect.

**Comment by charithjj** [comment]
@durandt Thank you very much.
Have you found any simple step by step guideline to setup ADFS ?

**Comment by durandt** [comment]
@charithjj I'm not the one in charge of the AD FS at my workplace so I did not do it myself, but here is a couple of links we found useful the first time we set it up: 
[OAuth2 authentication with ADFS 3.0](http://blog.scottlogic.com/2015/03/09/OAUTH2-Authentication-with-ADFS-3.0.html)
[Enable ADFS authentication for Mobile App](https://chitrarasan.wordpress.com/2014/07/16/enable-adfs-authentication-for-mobile-app/)

**Comment by rgccomtec** [comment]
Quick dirty solution.... 
https://42xyz.wordpress.com/2017/05/17/xamarin-and-oauth2-with-adfs/

**Comment by rgccomtec** [comment]
However I'd personally prefer to use ADAL as I mentioned here
https://42security.wordpress.com/2017/05/17/xamarin-oauth2-and-adfs-with-adal/

**Comment by charithjj** [comment]
Xamarin part seems straight forward. Unfortunately I have to setup the ADFS and I am bit confused with it.

@durandt @rgccomtec What do you guys think about Azure AD? Looks impressive to me. 
https://blog.xamarin.com/authenticate-xamarin-mobile-apps-using-azure-active-directory/

**Comment by rgccomtec** [comment]
You can use Azure AD but it may not be the best solution for many scenarios as Azure AD real features are charged on per uses basis... the free version is juts not enough for serious enterprise scenarios.

In case you want to go ahead with Azure AD then do either of following (assumption: users are in on premises AD):

Option 1: Move all on premises users to Azure AD using AADConnect including hash of hash of password sync feature
Option 2: Federate on premises AD with Azure AD... but then you will still have to setup ADFS (or similar entities like PING)

**Comment by charithjj** [comment]
@rgccomtec All the users are in On-Premises AD. I just want to authenticate and get the token for our mobile app. Probably I do need sync feature. I think we need only these basic features and looks like those features are free..

**Comment by rgccomtec** [comment]
Yes you can go ahead with AAD if password sync is fine with you. This will save you from setting up ADFS. However users won't get SSO (will be prompted for pwd at least once) and they will always need internet connectivity (no intranet only scenario).

**Comment by charithjj** [comment]
@rgccomtec I think that should be fine for me at least for now. Thank you for your help!

**Comment by durandt** [comment]
@charithjj If it hasn't changed since the last time i looked it up, Azure AD will not deliver sessions that can last longer than 90 days (it's not mobile friendly to have to type your password too often). 

Also, one big difference between Azure AD and on-prem AD FS is that you won't be able to get custom claims with Azure AD.

**Comment by charithjj** [comment]
@durandt @rgccomtec 
Looks like Azure AD is not the right solution for us. We have a large enterprise application and this new Xamarin ios mobile app just provides a few features. This enterprise application has already been deployed in many different organizations. I have a WebApi service for this mobile app that interfaces with enterprise application. So when we release this mobile app, they have to setup WebApi service. However, it turns out that many organizations don't keen on setting up Azure AD or even ADFS just to support this mobile app.

We actually don't need federated identity provider, because we only support on premise windows active directory.

So, I have to find a simpler solution which requires minimal network changes. My basic solution is to use IIS Basic Authentication and allow to user to enter and save domain username and password in appSettings. So I can send credentials in HttpWebRequest in a way that IIS BasicAuthentication can authenticate. When WebApi is publicly available, users can access the server from anywhere. Looks like this old school solution still works and provides everything we need. Do you see any issue with this approach?

Appreciate your help.

**Comment by rgccomtec** [comment]
My view… 

 

This will “work” but may not be aligned with security best practices. In my opinion, standards like OAuth2  did not come because something was “not possible”. Instead they came to do it better and widely accepted way. Basic Auth is considered to be the most insecure authentication mechanism on the web (no matter if we use SSL or not). The reason being… user’s credentials flow on the wire every time… with OAuth2 creds flow only once and thereafter tokens flow… so in essence it reduces the flow of creds over internet… another reason…submitting user’s AD credentials to every resource (app) is not a recommended security practice… AD creds should go to either an enterprise managed IdP or to DC only. Period! There is no need for creds to be bounced to every web service and web app and increase surface areas of attacks.

 

If ADFS (or similar) is not what orgs want then the need is to make them understand Cybersecurity best practice rather than offering insecure mechanisms. If price is a factor then ADFS is free and there are also free open source IdPs like thinktecture. 

 

 

-
Rahul Gangwar

**Comment by durandt** [comment]
@charithjj 
I don't think it's acceptable by todays industry standards. It may have been tolerated but by now everyone should create new applications using federated authentication, specially because it is so widespread (by providers like Google, Facebook, Twitter and many more). 

I won't go though all the security problems that your solution raises because it's already done, but be sure it's the main problem. Be aware that, if you get a user's credential, you get access to everything he has access to. At a company, that will be VPN, e-mail, web applications, mobiles applications and more. Whereas, if you get a user's session token, you can only access the system it was issued for. To me, that's the biggest difference between federated auth and username auth. 

Using username auth, you'll also create many problems/bugs that may be hard to solve if you don't get it right from the beginning: 
- Users will be logged out whenever they change their password
- If ONE client (you may have several apps) is not well implemented, the app, caching the old password, will try to communicate with your back-end, sometimes creating hundreds of calls to your APIs before the user logs out. This will block your user's account, if you have such policy. The entire account, just because one app misbehaved. 
- You'll have weird auth-related stuff happening because of backups (iCloud, etc)
- And many more problems the OAuth & Cie flows have fixed for you

Building username-auth today is building legacy code that will give you headaches. It's a bit more of a hassle to set-up federated auth for sure, but once it's set up you don't have to think about it anymore.

**Comment by charithjj** [comment]
@durandt @rgccomtec  You guys are on top of it. I completely agree with you.

We already have added the Xamarin part for AD FS. So we have both Basic and AD FS options available. I think we can provide both options so they can use what ever they want. If some organizations don't want to use AD FS, they can use Basic Auth with SSL.

I don't have much experience in security/network domain. When I tried to install ADFS in our organization, network engineers indicated that AD FS is normally for merging more than one active directories (looks like that's correct as per theory).  We have only one AD and that's the only way to authenticate users. If we don't federate more than one identify providers, using AD FS is apparently contradicts a bit with its basics. 

However, I agree with you. We need a more secure option. Thank you very much!

**Comment by newky2k** [comment]
resolved.
